# The Battle of Neighborhoods

### Introduction / Business Problem definition

In the US, one dense area is known to be highly different from the rest of the country in terms of innovation. The San Francisco Bay Area is indeed the most innovative place in the world, gathering more than 50% of all VC funds in the world, in a single, small region. 
The most important and famous parts of the Bay area include cities like San Francisco, Berkeley, San Jose, Palo Alto ... but the region is not limited to these territories and also include smaller cities a few kilometers away from the center of the Bay Area. 

Most of the highly successful tech companies were launched in this area and the highest paying jobs are located in that particular area. The San Francisco Bay Area is also well known for its diversity. In the area, it is often said that one is not from San Francisco if he lived there for more than a couple of years. 

Based on these few observations, we noticed an important problem that was created in the Bay Area: because of the average salary and the quality of life in this area, housing prices are completely disproportionate and for a new comer in the area, it can be a crazy riddle to solve when it comes to finding a place to live. 

In this analysis, I will get to know more about the Bay Area, all its locations and its main venues. Based on that, we will try to identify areas that can be similar (or not) to the city center of San Francisco, so that a new comer can easily identify where would be the best place for him to live in based on his preferences and salary. Throughout this analysis, we will try to understand why is Silicon Valley so different from the rest of the world and the rest of the Bay Area / California.

### Data

#### Data Sources

In order to achieve this analysis, I will need three types of information : 
- Information regarding the cities in The Bay Area, their postal codes, their associated coordinates (latitude and longitude). This information should be easily found on governmental websites : https://catalog.data.gov/dataset/bay-area-zip-codes/resource/6cacd1a1-6bff-4c7c-9094-49188ea29f85
- Associated venue to each of the different cities and neighborhoods, using Foursquare and its API.
- The San Francisco JSON file which provides a list of all Bay Area Boroughs, Neighborhoods, and their coordinates (latitude and longitude), if we cannot find it using the governmental websites information

### Methodology 

Methodology section which represents the main component of the report where you discuss and describe any exploratory data analysis that you did, any inferential statistical testing that you performed, if any, and what machine learnings were used and why.

Our methodology will be divided in different parts : 
1. Import packages and tools: In this section, we will import all tools and packages that we will require throughout the analysis. Those include but is not limited to: CSV, PD, NP, REQUESTS, KMEANS, JSON, GEOCODER, MATPLOTLIB, GEOCODER, FOLIUM, etc.
2. Data cleaning : Using the governmental website link that we found, let's import the CSV / HTML format in order to put it into a pandas dataframe. Once the required data is put into a pandas dataframe, we will be required to delete all the unecessary rows and columns, rename columns, add BLANK columns for latitude & longitude... The objective would be to have a table with columns for Neighborhood, ZIP code, Latitude and Longitude 
3. Use Geopy library: With Geopy library, we will be able to associate Coordinates (Latitude & Longitude) to the ZIP code. 
4. Use the FourSquare API: With Foursquare credentials, import all venues associated to the different locations. 
5. Once all the data is being collected and associated in one grouped table, run the Kmeans Machine Learning algorithm in order to determine what are the main clusters. Iterate based on the number of clusters required to get a good understanding of the data

#### Import & Data Cleaning

First let's import some key tools that will be useful throughout our analysis, these include but is not limited to numpy, pandas, jason ... 

In [1]:
import csv
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import numpy as np
import json # library to handle JSON files

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# import k-means from clustering stage
from sklearn.cluster import KMeans

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# Install geocoder
!pip install geocoder
import geocoder
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

!pip install folium
import folium # map rendering library

print('Libraries imported.')

     |████████████████████████████████| 98 kB 3.3 MB/s eta 0:00:011
     |████████████████████████████████| 93 kB 3.3 MB/s  eta 0:00:01
Libraries imported.


In [2]:
df = pd.read_csv ('https://data.sfgov.org/api/views/f9wk-m4qb/rows.csv?accessType=DOWNLOAD')   
df.head(10)

PO_NAME                                           the_geom    ZIP  \
0             NAPA  MULTIPOLYGON (((-122.10329200180091 38.5132829...  94558   
1        FAIRFIELD  MULTIPOLYGON (((-121.947475002335 38.301511000...  94533   
2            DIXON  MULTIPOLYGON (((-121.65335500334429 38.3133870...  95620   
3           SONOMA  MULTIPOLYGON (((-122.406843003057 38.155681999...  95476   
4             NAPA  MULTIPOLYGON (((-122.29368500225117 38.1552379...  94559   
5         PETALUMA  MULTIPOLYGON (((-122.45766900253919 38.1168949...  94954   
6        RIO VISTA  MULTIPOLYGON (((-121.8624620022998 38.06602999...  94571   
7       TRAVIS AFB  MULTIPOLYGON (((-121.89653900297888 38.2865679...  94535   
8  AMERICAN CANYON  MULTIPOLYGON (((-122.20418700285576 38.2096949...  94503   
9           NOVATO  MULTIPOLYGON (((-122.48655900081091 38.1005269...  94949   

  STATE        Area__       Length__  
0    CA  1.231326e+10  995176.225313  
1    CA  9.917861e+08  200772.556587  
2    CA  7.236950e+09  441860.201400  
3    CA  3.001414e+09  311318.546326  
4    CA  1.194302e+09  359104.646602  
5    CA  2.006544e+09  267474.490552  
6    CA  4.454446e+09  492056.752411  
7    CA  3.029397e+08   95232.008421  
8    CA  6.931341e+08  136394.695137  
9    CA  4.316054e+08  119395.672078

Rename PO_NAME column with a new name : "Neighborhood"

In [3]:
df = df.rename(columns={'PO_NAME': 'Neighborhood'})

Drop all unecessary columns, in order to get a better view of our table

In [4]:
neighborhoods = df.drop(['the_geom','STATE','Area__','Length__'], axis=1)
neighborhoods

Neighborhood    ZIP
0                   NAPA  94558
1              FAIRFIELD  94533
2                  DIXON  95620
3                 SONOMA  95476
4                   NAPA  94559
5               PETALUMA  94954
6              RIO VISTA  94571
7             TRAVIS AFB  94535
8        AMERICAN CANYON  94503
9                 NOVATO  94949
10                NOVATO  94945
11         BIRDS LANDING  94512
12               VALLEJO  94591
13               BENICIA  94510
14               VALLEJO  94592
15               VALLEJO  94589
16                NOVATO  94947
17               VALLEJO  94590
18               NICASIO  94946
19                OAKLEY  94561
20              CROCKETT  94525
21            PORT COSTA  94569
22           SUISUN CITY  94585
23         SAN FRANCISCO  94103
24             PITTSBURG  94565
25            SAN RAFAEL  94903
26               CONCORD  94520
27                 RODEO  94572
28              MARTINEZ  94553
29              HERCULES  94547
30          SAN GERONIMO  94963
31             LAGUNITAS  94938
32               ALAMEDA  94502
33               ANTIOCH  94509
34           SAN ANSELMO  94960
35             BRENTWOOD  94513
36         SAN FRANCISCO  94109
37               CONCORD  94521
38               FAIRFAX  94930
39              WOODACRE  94973
40         FOREST KNOLLS  94933
41          WALNUT CREEK  94598
42                PINOLE  94564
43              RICHMOND  94801
44               CONCORD  94519
45             SAN PABLO  94806
46            SAN RAFAEL  94901
47               ANTIOCH  94531
48           EL SOBRANTE  94803
49               OAKLAND  94601
50         PLEASANT HILL  94523
51               CONCORD  94518
52             GREENBRAE  94904
53         SAN FRANCISCO  94115
54             LAFAYETTE  94549
55               CLAYTON  94517
56              RICHMOND  94805
57              RICHMOND  94804
58              LARKSPUR  94939
59           SAN QUENTIN  94964
60            EL CERRITO  94530
61          CORTE MADERA  94925
62          WALNUT CREEK  94596
63              BERKELEY  94708
64         SAN FRANCISCO  94105
65           MILL VALLEY  94941
66                ORINDA  94563
67              BERKELEY  94720
68              BERKELEY  94707
69                 BYRON  94514
70         STINSON BEACH  94970
71                ALBANY  94706
72              BERKELEY  94710
73         SAN FRANCISCO  94104
74          WALNUT CREEK  94595
75              BERKELEY  94709
76              BERKELEY  94703
77              BERKELEY  94704
78                 ALAMO  94507
79              BERKELEY  94702
80             SAUSALITO  94965
81                MORAGA  94556
82     BELVEDERE TIBURON  94920
83         SAN FRANCISCO  94118
84              BERKELEY  94705
85               OAKLAND  94611
86               OAKLAND  94618
87               OAKLAND  94609
88             LIVERMORE  94550
89            EMERYVILLE  94608
90                DIABLO  94528
91              DANVILLE  94526
92              DANVILLE  94506
93         SAN FRANCISCO  94130
94               OAKLAND  94607
95         SAN FRANCISCO  94123
96               OAKLAND  94610
97             SAN RAMON  94583
98               OAKLAND  94602
99               OAKLAND  94612
100        CASTRO VALLEY  94546
101        SAN FRANCISCO  94133
102        SAN FRANCISCO  94129
103              OAKLAND  94606
104        SAN FRANCISCO  94111
105              OAKLAND  94619
106        SAN FRANCISCO  94121
107        SAN FRANCISCO  94102
108        CASTRO VALLEY  94552
109              ALAMEDA  94501
110        SAN FRANCISCO  94108
111              OAKLAND  94605
112              OAKLAND  94613
113        SAN FRANCISCO  94117
114        SAN FRANCISCO  94122
115              OAKLAND  94621
116        SAN FRANCISCO  94114
117        SAN FRANCISCO  94107
118        SAN FRANCISCO  94110
119           PLEASANTON  94588
120        SAN FRANCISCO  94131
121              OAKLAND  94603
122        SAN FRANCISCO  94116
123        SAN FRANCISCO  94124
124        S

In [5]:
print('The dataframe has {} neighborhoods (not yet compiled).'.format(
        neighborhoods.shape[0])
    )

The dataframe has 187 neighborhoods (not yet compiled).


In [6]:
print('The dataframe has the following number of UNIQUE neighborhoods:')
neighborhoods['Neighborhood'].nunique()

The dataframe has the following number of UNIQUE neighborhoods:


97

In [7]:
print('The dataframe has the following number of UNIQUE ZIP codes:')
neighborhoods['ZIP'].nunique()

The dataframe has the following number of UNIQUE ZIP codes:


187

Thanks to this analysis, we can easily notice that our dataset has 187 rows and 2 columns : Neighborhood & ZIP. In the Neighborhood column we have several repetitions (e.g. NOVATO rows 9 & 10) because we have only 97 unique neighborhoods in the Bay Area. We can also notice that we have 187 ZIP codes. 
Thus, we can conclude by saying that one Neighborhood has several ZIP codes associated. This information should be kept in mind when we will import the different information. We should rather focus on importing data based on the ZIP codes rather than based on the name of the Neighborhood.


Our database is now completely cleaned, we can start playing a little bit with Geopy, running some analysis on our data and all different functionalities. 

#### Use geopy library to get the latitude and longitude values of New York City.
In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>sf_explorer</em>, as shown below.


In [8]:
address = 'San Fransisco, CA'

geolocator = Nominatim(user_agent="sf_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of San Francisco are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of San Francisco are -0.8965048, -78.6549013.


Defining longitude and latitude for each different postal code using geocoder

In [9]:
neighborhoods['Latitude'] = None
neighborhoods['Longitude'] = None
neighborhoods

Neighborhood    ZIP Latitude Longitude
0                   NAPA  94558     None      None
1              FAIRFIELD  94533     None      None
2                  DIXON  95620     None      None
3                 SONOMA  95476     None      None
4                   NAPA  94559     None      None
5               PETALUMA  94954     None      None
6              RIO VISTA  94571     None      None
7             TRAVIS AFB  94535     None      None
8        AMERICAN CANYON  94503     None      None
9                 NOVATO  94949     None      None
10                NOVATO  94945     None      None
11         BIRDS LANDING  94512     None      None
12               VALLEJO  94591     None      None
13               BENICIA  94510     None      None
14               VALLEJO  94592     None      None
15               VALLEJO  94589     None      None
16                NOVATO  94947     None      None
17               VALLEJO  94590     None      None
18               NICASIO  94946     None      None
19                OAKLEY  94561     None      None
20              CROCKETT  94525     None      None
21            PORT COSTA  94569     None      None
22           SUISUN CITY  94585     None      None
23         SAN FRANCISCO  94103     None      None
24             PITTSBURG  94565     None      None
25            SAN RAFAEL  94903     None      None
26               CONCORD  94520     None      None
27                 RODEO  94572     None      None
28              MARTINEZ  94553     None      None
29              HERCULES  94547     None      None
30          SAN GERONIMO  94963     None      None
31             LAGUNITAS  94938     None      None
32               ALAMEDA  94502     None      None
33               ANTIOCH  94509     None      None
34           SAN ANSELMO  94960     None      None
35             BRENTWOOD  94513     None      None
36         SAN FRANCISCO  94109     None      None
37               CONCORD  94521     None      None
38               FAIRFAX  94930     None      None
39              WOODACRE  94973     None      None
40         FOREST KNOLLS  94933     None      None
41          WALNUT CREEK  94598     None      None
42                PINOLE  94564     None      None
43              RICHMOND  94801     None      None
44               CONCORD  94519     None      None
45             SAN PABLO  94806     None      None
46            SAN RAFAEL  94901     None      None
47               ANTIOCH  94531     None      None
48           EL SOBRANTE  94803     None      None
49               OAKLAND  94601     None      None
50         PLEASANT HILL  94523     None      None
51               CONCORD  94518     None      None
52             GREENBRAE  94904     None      None
53         SAN FRANCISCO  94115     None      None
54             LAFAYETTE  94549     None      None
55               CLAYTON  94517     None      None
56              RICHMOND  94805     None      None
57              RICHMOND  94804     None      None
58              LARKSPUR  94939     None      None
59           SAN QUENTIN  94964     None      None
60            EL CERRITO  94530     None      None
61          CORTE MADERA  94925     None      None
62          WALNUT CREEK  94596     None      None
63              BERKELEY  94708     None      None
64         SAN FRANCISCO  94105     None      None
65           MILL VALLEY  94941     None      None
66                ORINDA  94563     None      None
67              BERKELEY  94720     None      None
68              BERKELEY  94707     None      None
69                 BYRON  94514     None      None
70         STINSON BEACH  94970     None      None
71                ALBANY  94706     None      None
72              BERKELEY  94710     None      None
73         SAN FRANCISCO  94104     None      None
74          WALNUT CREEK  94595     None      None
75              BERKELEY  94709     None      None
76              BERKELEY  94703     None      None
77              BERKELEY  94704   

In [10]:
for i, postal_code in enumerate(neighborhoods['ZIP']):
    lat_lng_coords = None
    
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, CA'.format(postal_code))
        lat_lng_coords = g.latlng
    
    if lat_lng_coords:
        latitude = lat_lng_coords[0]
        longitude = lat_lng_coords[1]
    
    neighborhoods.loc[i, 'Latitude'] = latitude
    neighborhoods.loc[i, 'Longitude'] = longitude

neighborhoods

Neighborhood    ZIP Latitude Longitude
0                   NAPA  94558  38.3282  -122.328
1              FAIRFIELD  94533  38.2532  -122.046
2                  DIXON  95620  38.4476  -121.826
3                 SONOMA  95476  38.3124  -122.484
4                   NAPA  94559  38.3013   -122.29
5               PETALUMA  94954  38.2648  -122.641
6              RIO VISTA  94571  38.1593  -121.695
7             TRAVIS AFB  94535    38.28  -121.938
8        AMERICAN CANYON  94503  38.1627   -122.26
9                 NOVATO  94949  38.0698  -122.537
10                NOVATO  94945  38.1111   -122.57
11         BIRDS LANDING  94512  38.1316  -121.846
12               VALLEJO  94591  38.1065  -122.222
13               BENICIA  94510  38.0545  -122.157
14               VALLEJO  94592  38.0959  -122.272
15               VALLEJO  94589  38.1499  -122.261
16                NOVATO  94947  38.0993  -122.577
17               VALLEJO  94590  38.1072  -122.253
18               NICASIO  94946  38.0454  -122.648
19                OAKLEY  94561  37.9934  -121.718
20              CROCKETT  94525  38.0524   -122.22
21            PORT COSTA  94569  38.0456  -122.185
22           SUISUN CITY  94585  38.2453  -121.998
23         SAN FRANCISCO  94103   37.772  -122.409
24             PITTSBURG  94565  38.0282  -121.885
25            SAN RAFAEL  94903  38.0124  -122.543
26               CONCORD  94520  37.9735  -122.037
27                 RODEO  94572  38.0332  -122.266
28              MARTINEZ  94553  38.0139  -122.134
29              HERCULES  94547   38.012  -122.284
30          SAN GERONIMO  94963  38.0144  -122.668
31             LAGUNITAS  94938  38.0128    -122.7
32               ALAMEDA  94502  37.7341  -122.241
33               ANTIOCH  94509  38.0095  -121.812
34           SAN ANSELMO  94960  37.9767  -122.562
35             BRENTWOOD  94513  37.9352  -121.696
36         SAN FRANCISCO  94109  37.7901  -122.421
37               CONCORD  94521  37.9611  -121.996
38               FAIRFAX  94930  37.9857   -122.59
39              WOODACRE  94973   38.007  -122.632
40         FOREST KNOLLS  94933  38.0141  -122.684
41          WALNUT CREEK  94598  37.9133  -122.021
42                PINOLE  94564  38.0042  -122.295
43              RICHMOND  94801  37.9377  -122.362
44               CONCORD  94519  37.9805  -122.023
45             SAN PABLO  94806  37.9619  -122.343
46            SAN RAFAEL  94901  37.9702  -122.526
47               ANTIOCH  94531  37.9735  -121.769
48           EL SOBRANTE  94803  37.9745  -122.295
49               OAKLAND  94601   37.778  -122.222
50         PLEASANT HILL  94523  37.9575  -122.061
51               CONCORD  94518  37.9476  -122.033
52             GREENBRAE  94904  37.9484  -122.523
53         SAN FRANCISCO  94115  37.7849  -122.435
54             LAFAYETTE  94549  37.8924  -122.117
55               CLAYTON  94517  37.9353  -121.932
56              RICHMOND  94805  37.9383  -122.329
57              RICHMOND  94804  37.9298  -122.344
58              LARKSPUR  94939  37.9341  -122.535
59           SAN QUENTIN  94964  37.9414  -122.484
60            EL CERRITO  94530  37.9052  -122.301
61          CORTE MADERA  94925  37.9254  -122.523
62          WALNUT CREEK  94596  37.8995  -122.063
63              BERKELEY  94708  37.8868  -122.257
64         SAN FRANCISCO  94105  37.7897  -122.396
65           MILL VALLEY  94941  37.8983  -122.536
66                ORINDA  94563  37.8825  -122.185
67              BERKELEY  94720  37.8731  -122.258
68              BERKELEY  94707  37.8928  -122.278
69                 BYRON  94514  37.8702   -121.64
70         STINSON BEACH  94970  37.9009   -122.64
71                ALBANY  94706  37.8912  -122.296
72              BERKELEY  94710  37.8734  -122.298
73         SAN FRANCISCO  94104  37.7911  -122.402
74          WALNUT CREEK  94595  37.8888  -122.077
75              BERKELEY  94709  37.8785  -122.268
76              BERKELEY  94703   37.854  -122.273
77              BERKELEY  94704  3

Printing Bay-Area Map

In [11]:
map_bay_area = folium.Map(location=[latitude, longitude], zoom_start=8)

for lat, lng, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=4,
        popup=label,
        color='red',
        fill=True,
        parse_html=False).add_to(map_bay_area)
    
map_bay_area

We managed to get a great map of the Bay area and all the key cities and neighborhoods composing the region. We will now use Foursquare in order to move further with our analysis. 

### Foursquare and importing credentials

In [12]:
CLIENT_ID = 'K23DSBS00GSPJIUSF3VQGCCLLLHIOUSJ0244UPB41GFV4DZL' 
CLIENT_SECRET = 'OJ2S52HI3NQGDZL2IM2JZGX11OBYEG31CKBQZXWV2V3TN4RA'
VERSION = '20201207'
LIMIT = 100

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)
print('VERSION:' + VERSION)

Your credentails:
CLIENT_ID: K23DSBS00GSPJIUSF3VQGCCLLLHIOUSJ0244UPB41GFV4DZL
CLIENT_SECRET:OJ2S52HI3NQGDZL2IM2JZGX11OBYEG31CKBQZXWV2V3TN4RA
VERSION:20201207


Let's get all the different venue for each neighborhood by first defining the function that we will use ! 

In [13]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']


In [14]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

For each of the neighborhood and borough, let's now run the above function to get all venues in the Bay Area


In [15]:
bay_area_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                )

NAPA
FAIRFIELD
DIXON
SONOMA
NAPA
PETALUMA
RIO VISTA
TRAVIS AFB
AMERICAN CANYON
NOVATO
NOVATO
BIRDS LANDING
VALLEJO
BENICIA
VALLEJO
VALLEJO
NOVATO
VALLEJO
NICASIO
OAKLEY
CROCKETT
PORT COSTA
SUISUN CITY
SAN FRANCISCO
PITTSBURG
SAN RAFAEL
CONCORD
RODEO
MARTINEZ
HERCULES
SAN GERONIMO
LAGUNITAS
ALAMEDA
ANTIOCH
SAN ANSELMO
BRENTWOOD
SAN FRANCISCO
CONCORD
FAIRFAX
WOODACRE
FOREST KNOLLS
WALNUT CREEK
PINOLE
RICHMOND
CONCORD
SAN PABLO
SAN RAFAEL
ANTIOCH
EL SOBRANTE
OAKLAND
PLEASANT HILL
CONCORD
GREENBRAE
SAN FRANCISCO
LAFAYETTE
CLAYTON
RICHMOND
RICHMOND
LARKSPUR
SAN QUENTIN
EL CERRITO
CORTE MADERA
WALNUT CREEK
BERKELEY
SAN FRANCISCO
MILL VALLEY
ORINDA
BERKELEY
BERKELEY
BYRON
STINSON BEACH
ALBANY
BERKELEY
SAN FRANCISCO
WALNUT CREEK
BERKELEY
BERKELEY
BERKELEY
ALAMO
BERKELEY
SAUSALITO
MORAGA
BELVEDERE TIBURON
SAN FRANCISCO
BERKELEY
OAKLAND
OAKLAND
OAKLAND
LIVERMORE
EMERYVILLE
DIABLO
DANVILLE
DANVILLE
SAN FRANCISCO
OAKLAND
SAN FRANCISCO
OAKLAND
SAN RAMON
OAKLAND
OAKLAND
CASTRO VALLEY
SAN FRANCISCO
S

In [16]:
print(bay_area_venues.shape)
bay_area_venues.head()

(5121, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         NAPA              38.328225             -122.328115   
1    FAIRFIELD              38.253238             -122.046009   
2    FAIRFIELD              38.253238             -122.046009   
3    FAIRFIELD              38.253238             -122.046009   
4    FAIRFIELD              38.253238             -122.046009   

                         Venue  Venue Latitude  Venue Longitude  \
0                  Alston Park       38.327588      -122.332593   
1                    Chez Soul       38.249759      -122.046145   
2                 Joe's Buffet       38.249450      -122.044118   
3                 China Palace       38.249492      -122.045125   
4  Rustwater Kitchen & Taproom       38.249426      -122.043313   

                    Venue Category  
0                            Trail  
1  Southern / Soul Food Restaurant  
2                   Sandwich Place  
3               Chinese Restaurant  
4                         Beer Bar

Let's see how many venues appeared for each neighborhood

In [17]:
bay_area_venues.groupby('Neighborhood').count()


Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
ALAMEDA                                 91                      91     91   
ALAMO                                    2                       2      2   
ALBANY                                  48                      48     48   
ALVISO                                   5                       5      5   
AMERICAN CANYON                          3                       3      3   
ANTIOCH                                  7                       7      7   
ATHERTON                                 5                       5      5   
BELMONT                                 50                      50     50   
BELVEDERE TIBURON                       16                      16     16   
BENICIA                                 28                      28     28   
BERKELEY                               380                     380    380   
BRENTWOOD                               36                      36     36   
BRISBANE                                28                      28     28   
BURLINGAME                              20                      20     20   
BYRON                                    4                       4      4   
CASTRO VALLEY                           47                      47     47   
CLAYTON                                  1                       1      1   
CONCORD                                 95                      95     95   
CORTE MADERA                             8                       8      8   
CROCKETT                                 8                       8      8   
DALY CITY                               38                      38     38   
DANVILLE                                89                      89     89   
DIABLO                                   2                       2      2   
DIXON                                    9                       9      9   
DUBLIN                                  31                      31     31   
EL CERRITO                              28                      28     28   
EL SOBRANTE                             13                      13     13   
EMERYVILLE                              19                      19     19   
FAIRFAX                                 34                      34     34   
FAIRFIELD                                9                       9      9   
FOREST KNOLLS                            6                       6      6   
FREMONT                                103                     103    103   
GREENBRAE                               26                      26     26   
HALF MOON BAY                           52                      52     52   
HAYWARD                                 86                      86     86   
HERCULES                                 1                       1      1   
LAFAYETTE                               51                      51     51   
LAGUNITAS                                4                       4      4   
LARKSPUR                                31                      31     31   
LIVERMORE                               43                      43     43   
LOS ALTOS                               43                      43     43   
MARTINEZ                                18                      18     18   
MENLO PARK                              44                      44     44   
MILL VALLEY                             25                      25     25   
MILLBRAE                                24                      24     24   
MILPITAS                                96                      96     96   
MORAGA                                   4                       4      4   
MOSS BEACH                              10                      10     10   
MOUNT HAMILTON                           2                       2      2   
MOUNTAIN VIEW                           46                      46     46   
NAPA             

In [18]:
print('There are {} uniques categories.'.format(len(bay_area_venues['Venue Category'].unique())))

There are 374 uniques categories.


We can thus notice that there is a very large number of unique categories within the different venues related to the Bay Area (378)

We can now notice that we have a total of 97 neighborhood in our dataset, and each of these neighborhood have an associated number of venues. Let's now further analyze these venues by determining the kind of venue (Art, Restaurant, Monument, video, clubs, bar...) in each neighborhood : 

In [19]:
# one hot encoding
bay_area_onehot = pd.get_dummies(bay_area_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
bay_area_onehot['Neighborhood'] = bay_area_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [bay_area_onehot.columns[-1]] + list(bay_area_onehot.columns[:-1])
bay_area_onehot = bay_area_onehot[fixed_columns]

bay_area_onehot.head()

Yoga Studio  ATM  Acai House  Accessories Store  Acupuncturist  \
0            0    0           0                  0              0   
1            0    0           0                  0              0   
2            0    0           0                  0              0   
3            0    0           0                  0              0   
4            0    0           0                  0              0   

   Adult Boutique  Afghan Restaurant  African Restaurant  Airport  \
0               0                  0                   0        0   
1               0                  0                   0        0   
2               0                  0                   0        0   
3               0                  0                   0        0   
4               0                  0                   0        0   

   Alternative Healer  American Restaurant  Amphitheater  Andhra Restaurant  \
0                   0                    0             0                  0   
1                   0                    0             0                  0   
2                   0                    0             0                  0   
3                   0                    0             0                  0   
4                   0                    0             0                  0   

   Animal Shelter  Antique Shop  Arcade  Argentinian Restaurant  Art Gallery  \
0               0             0       0                       0            0   
1               0             0       0                       0            0   
2               0             0       0                       0            0   
3               0             0       0                       0            0   
4               0             0       0                       0            0   

   Art Museum  Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0           0                    0                     0                 0   
1           0                    0                     0                 0   
2           0                    0                     0                 0   
3           0                    0                     0                 0   
4           0                    0                     0                 0   

   Athletics & Sports  Auditorium  Auto Garage  Auto Workshop  \
0                   0           0            0              0   
1                   0           0            0              0   
2                   0           0            0              0   
3                   0           0            0              0   
4                   0           0            0              0   

   Automotive Shop  BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0                0          0           0       0     0    0               0   
1                0          0           0       0     0    0               0   
2                0          0           0       0     0    0               0   
3                0          0           0       0     0    0               0   
4                0          0           0       0     0    0               0   

   Basketball Court  Bath House  Bay  Beach  Bed & Breakfast  Beer Bar  \
0                 0           0    0      0                0         0   
1                 0           0    0      0                0         0   
2                 0           0    0      0                0         0   
3                 0           0    0      0                0         0   
4                 0           0    0      0                0         1   

   Beer Garden  Beer Store  Big Box Store  Bike Shop  Bike Trail  Bistro  \
0            0           0              0          0           0       0   
1            0           0              0          0           0       0   
2            0           0              0          0           0       0   
3            0           0              0          0           0       0   
4            0           0              0          

In [20]:
bay_area_onehot.shape

(5121, 374)

The size of the new dataframe is coherent as we can see that we have 5,178 rows corresponding to the 5,178 venues and 378 columns corresponding to the 378 unique categories. 

In [21]:
bay_area_grouped = bay_area_onehot.groupby('Neighborhood').mean().reset_index()
bay_area_grouped

Neighborhood  Yoga Studio       ATM  Acai House  Accessories Store  \
0               ALAMEDA     0.000000  0.021978     0.00000           0.010989   
1                 ALAMO     0.000000  0.000000     0.00000           0.000000   
2                ALBANY     0.000000  0.000000     0.00000           0.000000   
3                ALVISO     0.000000  0.000000     0.00000           0.000000   
4       AMERICAN CANYON     0.000000  0.333333     0.00000           0.000000   
5               ANTIOCH     0.000000  0.000000     0.00000           0.000000   
6              ATHERTON     0.000000  0.000000     0.00000           0.000000   
7               BELMONT     0.000000  0.020000     0.00000           0.000000   
8     BELVEDERE TIBURON     0.000000  0.000000     0.00000           0.000000   
9               BENICIA     0.000000  0.000000     0.00000           0.000000   
10             BERKELEY     0.007895  0.000000     0.00000           0.005263   
11            BRENTWOOD     0.000000  0.000000     0.00000           0.000000   
12             BRISBANE     0.000000  0.000000     0.00000           0.000000   
13           BURLINGAME     0.000000  0.000000     0.00000           0.000000   
14                BYRON     0.000000  0.000000     0.00000           0.000000   
15        CASTRO VALLEY     0.000000  0.000000     0.00000           0.000000   
16              CLAYTON     0.000000  0.000000     0.00000           0.000000   
17              CONCORD     0.000000  0.010526     0.00000           0.000000   
18         CORTE MADERA     0.000000  0.000000     0.00000           0.000000   
19             CROCKETT     0.000000  0.000000     0.00000           0.000000   
20            DALY CITY     0.000000  0.000000     0.00000           0.000000   
21             DANVILLE     0.011236  0.000000     0.00000           0.000000   
22               DIABLO     0.000000  0.000000     0.00000           0.000000   
23                DIXON     0.000000  0.000000     0.00000           0.000000   
24               DUBLIN     0.000000  0.000000     0.00000           0.000000   
25           EL CERRITO     0.000000  0.000000     0.00000           0.000000   
26          EL SOBRANTE     0.000000  0.000000     0.00000           0.000000   
27           EMERYVILLE     0.000000  0.000000     0.00000           0.000000   
28              FAIRFAX     0.029412  0.000000     0.00000           0.000000   
29            FAIRFIELD     0.000000  0.000000     0.00000           0.000000   
30        FOREST KNOLLS     0.000000  0.000000     0.00000           0.000000   
31              FREMONT     0.000000  0.000000     0.00000           0.000000   
32            GREENBRAE     0.000000  0.000000     0.00000           0.000000   
33        HALF MOON BAY     0.000000  0.000000     0.00000           0.000000   
34              HAYWARD     0.011628  0.011628     0.00000           0.000000   
35             HERCULES     0.000000  0.000000     0.00000           0.000000   
36            LAFAYETTE     0.000000  0.019608     0.00000           0.000000   
37            LAGUNITAS     0.000000  0.000000     0.00000           0.000000   
38             LARKSPUR     0.000000  0.000000     0.00000           0.000000   
39            LIVERMORE     0.000000  0.000000     0.00000           0.000000   
40            LOS ALTOS     0.000000  0.023256     0.00000           0.000000   
41             MARTINEZ     0.000000  0.000000     0.00000           0.000000   
42           MENLO PARK     0.000000  0.000000     0.00000           0.000000   
43          MILL VALLEY     0.040000  0.000000     0.00000           0.000000   
44             MILLBRAE     0.000000  0.000000     0.00000           0.000000   
45             MILPITAS     0.000000  0.000000     0.00000           0.041667   
46               MORAGA     0.000000  0.000000     0.00000           0.000000   
47           MOSS BEACH     0.100000  0.000000     0.00000           0.000000   
48       MOUNT HAMILTON     0.000000  0.00

In [22]:
bay_area_grouped.shape

(94, 374)

In this new dataframe, we can notice that we still have the previous 378 columns. But this time we only have 94 rows, this can be explained by the grouping of identical neighborhoods with eachother.
Let's now take a look at the TOP 5 venues for each neighborhood

In [23]:
num_top_venues = 5

for hood in bay_area_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = bay_area_grouped[bay_area_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----ALAMEDA----
                 venue  freq
0          Coffee Shop  0.03
1   Mexican Restaurant  0.03
2  American Restaurant  0.03
3   Italian Restaurant  0.03
4           Nail Salon  0.03


----ALAMO----
               venue  freq
0        Art Gallery   0.5
1       Home Service   0.5
2        Yoga Studio   0.0
3          Nightclub   0.0
4  Other Repair Shop   0.0


----ALBANY----
               venue  freq
0    Thai Restaurant  0.06
1     Breakfast Spot  0.04
2                Bar  0.04
3       Liquor Store  0.04
4  French Restaurant  0.04


----ALVISO----
                venue  freq
0  Mexican Restaurant   0.6
1          Food Truck   0.2
2               Plaza   0.2
3         Yoga Studio   0.0
4           Newsstand   0.0


----AMERICAN CANYON----
          venue  freq
0           ATM  0.33
1  Tennis Court  0.33
2      Pharmacy  0.33
3     Newsstand  0.00
4  Optical Shop  0.00


----ANTIOCH----
                venue  freq
0          Playground  0.29
1                 Gym  0.14
2       

In [24]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [25]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = bay_area_grouped['Neighborhood']

for ind in np.arange(bay_area_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(bay_area_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0          ALAMEDA            Nail Salon    Italian Restaurant   
1            ALAMO           Art Gallery          Home Service   
2           ALBANY       Thai Restaurant                   Bar   
3           ALVISO    Mexican Restaurant                 Plaza   
4  AMERICAN CANYON                   ATM          Tennis Court   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop    Mexican Restaurant   American Restaurant   
1         Women's Store                  Farm  Ethiopian Restaurant   
2           Coffee Shop     French Restaurant                  Café   
3            Food Truck         Women's Store                  Farm   
4              Pharmacy         Women's Store                  Farm   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0       Thai Restaurant              Pharmacy          Dessert Shop   
1         Event Service           Event Space               Exhibit   
2             Pet Store  Gym / Fitness Center     Indian Restaurant   
3  Ethiopian Restaurant         Event Service           Event Space   
4  Ethiopian Restaurant         Event Service           Event Space   

  9th Most Common Venue   10th Most Common Venue  
0      Asian Restaurant  New American Restaurant  
1            Eye Doctor              Fabric Shop  
2        Breakfast Spot             Liquor Store  
3               Exhibit               Eye Doctor  
4               Exhibit               Eye Doctor

In [26]:
# set number of clusters
k = 50

bay_area_grouped_clustering = bay_area_grouped.drop('Neighborhood', axis=1)

# run k-means clustering
kmeans = KMeans(n_clusters=k, random_state=1).fit(bay_area_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:50] 

array([22, 16, 22, 10, 25, 27,  0, 19, 30, 22, 22,  1, 46, 28, 14, 22,  2,
       22, 29, 21, 19, 22,  9, 38, 19, 22, 31, 41, 22, 24, 26, 19, 22, 22,
        1,  7, 22,  5, 22, 22, 22, 44, 22, 22, 22, 39, 18, 37, 12, 22],
      dtype=int32)

In [27]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

bay_area_merged = neighborhoods
bay_area_merged = bay_area_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
bay_area_merged.dropna(inplace=True)


In [28]:
print(bay_area_merged.shape)
bay_area_merged['Cluster Labels'] = bay_area_merged['Cluster Labels'].astype(int)
bay_area_merged

(184, 15)


Neighborhood    ZIP Latitude Longitude  Cluster Labels  \
0                   NAPA  94558  38.3282  -122.328              22   
1              FAIRFIELD  94533  38.2532  -122.046              24   
2                  DIXON  95620  38.4476  -121.826              38   
3                 SONOMA  95476  38.3124  -122.484              48   
4                   NAPA  94559  38.3013   -122.29              22   
5               PETALUMA  94954  38.2648  -122.641              34   
6              RIO VISTA  94571  38.1593  -121.695              43   
7             TRAVIS AFB  94535    38.28  -121.938              20   
8        AMERICAN CANYON  94503  38.1627   -122.26              25   
9                 NOVATO  94949  38.0698  -122.537               1   
10                NOVATO  94945  38.1111   -122.57               1   
12               VALLEJO  94591  38.1065  -122.222              22   
13               BENICIA  94510  38.0545  -122.157              22   
14               VALLEJO  94592  38.0959  -122.272              22   
15               VALLEJO  94589  38.1499  -122.261              22   
16                NOVATO  94947  38.0993  -122.577               1   
17               VALLEJO  94590  38.1072  -122.253              22   
19                OAKLEY  94561  37.9934  -121.718               8   
20              CROCKETT  94525  38.0524   -122.22              21   
21            PORT COSTA  94569  38.0456  -122.185              36   
23         SAN FRANCISCO  94103   37.772  -122.409              22   
24             PITTSBURG  94565  38.0282  -121.885              45   
25            SAN RAFAEL  94903  38.0124  -122.543              22   
26               CONCORD  94520  37.9735  -122.037              22   
27                 RODEO  94572  38.0332  -122.266              49   
28              MARTINEZ  94553  38.0139  -122.134              44   
29              HERCULES  94547   38.012  -122.284               7   
30          SAN GERONIMO  94963  38.0144  -122.668               6   
31             LAGUNITAS  94938  38.0128    -122.7               5   
32               ALAMEDA  94502  37.7341  -122.241              22   
33               ANTIOCH  94509  38.0095  -121.812              27   
34           SAN ANSELMO  94960  37.9767  -122.562              22   
35             BRENTWOOD  94513  37.9352  -121.696               1   
36         SAN FRANCISCO  94109  37.7901  -122.421              22   
37               CONCORD  94521  37.9611  -121.996              22   
38               FAIRFAX  94930  37.9857   -122.59              22   
39              WOODACRE  94973   38.007  -122.632               4   
40         FOREST KNOLLS  94933  38.0141  -122.684              26   
41          WALNUT CREEK  94598  37.9133  -122.021              22   
42                PINOLE  94564  38.0042  -122.295              33   
43              RICHMOND  94801  37.9377  -122.362              40   
44               CONCORD  94519  37.9805  -122.023              22   
45             SAN PABLO  94806  37.9619  -122.343               1   
46            SAN RAFAEL  94901  37.9702  -122.526              22   
47               ANTIOCH  94531  37.9735  -121.769              27   
48           EL SOBRANTE  94803  37.9745  -122.295              31   
49               OAKLAND  94601   37.778  -122.222              22   
50         PLEASANT HILL  94523  37.9575  -122.061              22   
51               CONCORD  94518  37.9476  -122.033              22   
52             GREENBRAE  94904  37.9484  -122.523              22   
53         SAN FRANCISCO  94115  37.7849  -122.435              22   
54             LAFAYETTE  94549  37.8924  -122.117              22   
55               CLAYTON  94517  37.9353  -121.932               2   
56              RICHMOND  94805  37.9383  -122.329              40   
57              RICHMOND  94804  37.9298  -122.344              40   
58              LARKSPUR  94939  37.9341  -122.535              22   
59           SAN QUENT

In [29]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(bay_area_merged['Latitude'], bay_area_merged['Longitude'], bay_area_merged['Neighborhood'], bay_area_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Let's now have list of all different clusters : I only put the first 5 clusters, but we could easily focus on the other by changing the figure associated. 

In [37]:
bay_area_merged.loc[bay_area_merged['Cluster Labels'] == 1, bay_area_merged.columns[[1] + list(range(5, bay_area_merged.shape[1]))]]


ZIP 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
9    94949    Mexican Restaurant           Pizza Place                Bakery   
10   94945    Mexican Restaurant           Pizza Place                Bakery   
16   94947    Mexican Restaurant           Pizza Place                Bakery   
35   94513           Pizza Place    Mexican Restaurant   American Restaurant   
45   94806    Mexican Restaurant           Supermarket           Pizza Place   
136  94541    Mexican Restaurant           Pizza Place                   Bar   
138  94542    Mexican Restaurant           Pizza Place                   Bar   
139  94544    Mexican Restaurant           Pizza Place                   Bar   
141  94545    Mexican Restaurant           Pizza Place                   Bar   
143  94080    Mexican Restaurant           Coffee Shop   Japanese Restaurant   
164  94070    Mexican Restaurant           Pizza Place    Chinese Restaurant   

    4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
9                     Bar           Gas Station   Rental Car Location   
10                    Bar           Gas Station   Rental Car Location   
16                    Bar           Gas Station   Rental Car Location   
35         Sandwich Place                   Bar        Shipping Store   
45          Hot Dog Joint          Dessert Shop          Liquor Store   
136           Coffee Shop    Italian Restaurant    Chinese Restaurant   
138           Coffee Shop    Italian Restaurant    Chinese Restaurant   
139           Coffee Shop    Italian Restaurant    Chinese Restaurant   
141           Coffee Shop    Italian Restaurant    Chinese Restaurant   
143        Sandwich Place    Italian Restaurant                 Diner   
164                   Spa         Auto Workshop       Automotive Shop   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
9                     Gym        Sandwich Place    Italian Restaurant   
10                    Gym        Sandwich Place    Italian Restaurant   
16                    Gym        Sandwich Place    Italian Restaurant   
35            Gas Station                  Bank            Laundromat   
45        Thai Restaurant          Optical Shop                   ATM   
136      Sushi Restaurant                  Park          Cupcake Shop   
138      Sushi Restaurant                  Park          Cupcake Shop   
139      Sushi Restaurant                  Park          Cupcake Shop   
141      Sushi Restaurant                  Park          Cupcake Shop   
143           Gas Station   Rental Car Location       Thai Restaurant   
164      Tapas Restaurant           Gas Station      Sushi Restaurant   

    10th Most Common Venue  
9           Breakfast Spot  
10          Breakfast Spot  
16          Breakfast Spot  
35              Bagel Shop  
45     Fried Chicken Joint  
136         Ice Cream Shop  
138         Ice Cream Shop  
139         Ice Cream Shop  
141         Ice Cream Shop  
143     Chinese Restaurant  
164         Massage Studio

In [38]:
bay_area_merged.loc[bay_area_merged['Cluster Labels'] == 22, bay_area_merged.columns[[1] + list(range(5, bay_area_merged.shape[1]))]]


ZIP  1st Most Common Venue   2nd Most Common Venue  \
0    94558               Wine Bar      Mexican Restaurant   
4    94559               Wine Bar      Mexican Restaurant   
12   94591                   Park                  Bakery   
13   94510     Italian Restaurant             Pizza Place   
14   94592                   Park                  Bakery   
15   94589                   Park                  Bakery   
17   94590                   Park                  Bakery   
23   94103            Coffee Shop                    Park   
25   94903            Coffee Shop      Mexican Restaurant   
26   94520     Mexican Restaurant          Sandwich Place   
32   94502             Nail Salon      Italian Restaurant   
34   94960     Mexican Restaurant           Grocery Store   
36   94109            Coffee Shop                    Park   
37   94521     Mexican Restaurant          Sandwich Place   
38   94930                    Bar      Italian Restaurant   
41   94598            Coffee Shop             Pizza Place   
44   94519     Mexican Restaurant          Sandwich Place   
46   94901            Coffee Shop      Mexican Restaurant   
49   94601                   Café             Coffee Shop   
50   94523    Sporting Goods Shop          Shop & Service   
51   94518     Mexican Restaurant          Sandwich Place   
52   94904            Pizza Place                    Bank   
53   94115            Coffee Shop                    Park   
54   94549            Coffee Shop            Liquor Store   
58   94939    American Restaurant      Italian Restaurant   
60   94530  Vietnamese Restaurant       Mobile Phone Shop   
62   94596            Coffee Shop             Pizza Place   
63   94708            Coffee Shop                    Café   
64   94105            Coffee Shop                    Park   
65   94941            Pizza Place          Breakfast Spot   
66   94563           Optical Shop             Pizza Place   
67   94720            Coffee Shop                    Café   
68   94707            Coffee Shop                    Café   
71   94706        Thai Restaurant                     Bar   
72   94710            Coffee Shop                    Café   
73   94104            Coffee Shop                    Park   
74   94595            Coffee Shop             Pizza Place   
75   94709            Coffee Shop                    Café   
76   94703            Coffee Shop                    Café   
77   94704            Coffee Shop                    Café   
79   94702            Coffee Shop                    Café   
83   94118            Coffee Shop                    Park   
84   94705            Coffee Shop                    Café   
85   94611                   Café             Coffee Shop   
86   94618                   Café             Coffee Shop   
87   94609                   Café             Coffee Shop   
88   94550     Mexican Restaurant             Coffee Shop   
91   94526            Pizza Place          Sandwich Place   
92   94506            Pizza Place          Sandwich Place   
93   94130            Coffee Shop                    Park   
94   94607                   Café             Coffee Shop   
95   94123            Coffee Shop                    Park   
96   94610                   Café             Coffee Shop   
97   94583           Liquor Store             Coffee Shop   
98   94602                   Café             Coffee Shop   
99   94612                   Café             Coffee Shop   
100  94546         Ice Cream Shop                     Bar   
101  94133            Coffee Shop                    Park   
102  94129            Coffee Shop                    Park   
103  94606                   Café             Coffee Shop   
104  94111            Coffee Shop                    Park   
105  94619                   Café             Coffee Shop   
106  94121            Coffee Shop                    Park   
107  94102            Coffee Shop                    Park   
108  94552         Ice Cream Shop         

In [39]:
bay_area_merged.loc[bay_area_merged['Cluster Labels'] == 40, bay_area_merged.columns[[1] + list(range(5, bay_area_merged.shape[1]))]]


ZIP 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
43  94801    Mexican Restaurant              Pharmacy     Convenience Store   
56  94805    Mexican Restaurant              Pharmacy     Convenience Store   
57  94804    Mexican Restaurant              Pharmacy     Convenience Store   

   4th Most Common Venue 5th Most Common Venue   6th Most Common Venue  \
43           Coffee Shop   Fried Chicken Joint  Furniture / Home Store   
56           Coffee Shop   Fried Chicken Joint  Furniture / Home Store   
57           Coffee Shop   Fried Chicken Joint  Furniture / Home Store   

   7th Most Common Venue 8th Most Common Venue   9th Most Common Venue  \
43            Bagel Shop        Sandwich Place  Thrift / Vintage Store   
56            Bagel Shop        Sandwich Place  Thrift / Vintage Store   
57            Bagel Shop        Sandwich Place  Thrift / Vintage Store   

   10th Most Common Venue  
43     Chinese Restaurant  
56     Chinese Restaurant  
57     Chinese Restaurant

In [40]:
bay_area_merged.loc[bay_area_merged['Cluster Labels'] == 3, bay_area_merged.columns[[1] + list(range(5, bay_area_merged.shape[1]))]]


ZIP 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
144  94587                  Park         Women's Store                  Farm   

    4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
144  Ethiopian Restaurant         Event Service           Event Space   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
144               Exhibit            Eye Doctor           Fabric Shop   

    10th Most Common Venue  
144                Factory

In [41]:
bay_area_merged.loc[bay_area_merged['Cluster Labels'] == 4, bay_area_merged.columns[[1] + list(range(5, bay_area_merged.shape[1]))]]


ZIP 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
39  94973        Baseball Field         Women's Store        Farmers Market   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
39         Event Service           Event Space               Exhibit   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
39            Eye Doctor           Fabric Shop               Factory   

   10th Most Common Venue  
39     Falafel Restaurant